### PR TITLE
Add nuance around stretched clusters

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the multi-zone-cluster design docs say to keep all the nodes in a
single datacenter. This doesn't really reflect what we do in practice:
each zone in AWS/GCP/Azure/etc is a separate datacenter with decent
connectivity to the other zones in the same region. This commit adjusts
the docs to allow for this.

### CODE REVIEW

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 1]
```diff
@@ -230,24 +230,44 @@ The cluster will be resilient to the loss of any node as long as:
...
+zones within a single data center with each zone having its own independent
+power supply and other supporting infrastructure. You can also _stretch_ your
+cluster across nearby data centers as long as the network interconnection
+between each pair of data centers is good enough.
+
+[[high-availability-cluster-design-min-network-perf]] There is no specific
+minimum network performance required to run a healthy {es} cluster. In theory a
+cluster will work correctly even if the round-trip latency between nodes is
+several hundred milliseconds. In practice if your network is that slow then the
+cluster performance will be very poor. In addition, slow networks are often
```
Comment (mjmbischoff) [inline docs/reference/high-availability/cluster-design.asciidoc:262]:
```suggestion
cluster performance will likely be at unacceptable levels. In addition, slow networks are often
```

**Review by mjmbischoff — APPROVED** [review]
I think it's good, added some comments but mainly to add some field context. So fee free to ignore the literal suggestion and just adapt accordingly if the comment resonates.

Pinging @deepybee as he's a better word smith then I am :-)

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 2]
```diff
@@ -230,24 +230,44 @@ The cluster will be resilient to the loss of any node as long as:
...
+A few milliseconds of latency added to each round-trip can quickly accumulate
+into a noticeable performance penalty. {es} will automatically recover from a
+network partition as quickly as it can but your cluster may be partly
+unavailable during a partition and will need to spend time and resources to
+resynchronize any missing data and rebalance itself once a partition heals.
+
+If you have divided your cluster into zones then typically the network
+connections within each zone are of higher quality than the connections between
+the zones. You must make sure that the network connections between zones are of
+sufficiently high quality. You will see the best results by locating all your
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc:252]: I'd rather be slightly more vague here: it's not just about performance, reliability is also a big deal.

File: docs/reference/high-availability/cluster-design.asciidoc (hunk 1 above)
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc:262]: I started with something like that but then I figured we'd have to dive into what "unacceptable" means and how you'd determine what is or isn't acceptable. I saw someone running a very stretched cluster over satellite links once. Its performance was terrible in an absolute sense, and yet it was still acceptable to them. There's certainly a place for that sort of discussion but it's not here.

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 3]
```diff
@@ -230,24 +230,44 @@ The cluster will be resilient to the loss of any node as long as:
 [[high-availability-cluster-design-large-clusters]]
 === Resilience in larger clusters
 
-It is not unusual for nodes to share some common infrastructure, such as a power
-supply or network router. If so, you should plan for the failure of this
+It is not unusual for nodes to share some common infrastructure, such as network
+interconnects or a power supply. If so, you should plan for the failure of this
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc:234]: I'm not sure about this. I mean it's correct but it does make the sentence much more complicated. Is it worth the extra words? Do we need to clarify that nodes on the same physical host share infrastructure like power and network? Seems kinda obvious to me but this is a genuine question, I'm not the one on the front line for this kind of thing.

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 4]
```diff
@@ -230,24 +230,44 @@ The cluster will be resilient to the loss of any node as long as:
...
-high bandwidth. Connections between data centres typically do not meet these
-expectations. Although {es} will behave correctly on an unreliable or slow
-network, it will not necessarily behave optimally. It may take a considerable
-length of time for a cluster to fully recover from a network partition since it
-must resynchronize any missing data and rebalance the cluster once the
-partition heals. If you want your data to be available in multiple data centres,
-deploy a separate cluster in each data centre and use
-<<modules-cross-cluster-search,{ccs}>> or <<xpack-ccr,{ccr}>> to link the
+{es} expects its node-to-node connections to be reliable and have low latency
+and good bandwidth. Many of the tasks that {es} performs require multiple
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc]: "Enough bandwidth" feels awkward to me, how about "adequate bandwidth"? See b0fae802ca7b566666e5043c2f0b3d375895301f.

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 5]
```diff
@@ -230,24 +230,44 @@ The cluster will be resilient to the loss of any node as long as:
...
+round-trips between nodes. This means that a slow or unreliable interconnect
+may have a significant effect on the performance and stability of your cluster.
+A few milliseconds of latency added to each round-trip can quickly accumulate
+into a noticeable performance penalty. {es} will automatically recover from a
+network partition as quickly as it can but your cluster may be partly
+unavailable during a partition and will need to spend time and resources to
+resynchronize any missing data and rebalance itself once a partition heals.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc]: Good point, thanks. Added a sentence at the end of this paragraph about recovery time in b0fae802ca7b566666e5043c2f0b3d375895301f.

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 6]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
-expectations. Although {es} will behave correctly on an unreliable or slow
-network, it will not necessarily behave optimally. It may take a considerable
-length of time for a cluster to fully recover from a network partition since it
-must resynchronize any missing data and rebalance the cluster once the
-partition heals. If you want your data to be available in multiple data centres,
-deploy a separate cluster in each data centre and use
-<<modules-cross-cluster-search,{ccs}>> or <<xpack-ccr,{ccr}>> to link the
+{es} expects its node-to-node connections to be reliable and to have low
+latency and adequate bandwidth. Many of the tasks that {es} performs require
+multiple round-trips between nodes. This means that a slow or unreliable
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:241]:
```suggestion
{es} expects node-to-node connections to be reliable, have low latency, and
have adequate bandwidth. Many {es} tasks require multiple round-trips between
nodes. A slow or unreliable
```

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 7]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
+interconnect may have a significant effect on the performance and stability of
+your cluster. A few milliseconds of latency added to each round-trip can
+quickly accumulate into a noticeable performance penalty. {es} will
+automatically recover from a network partition as quickly as it can but your
+cluster may be partly unavailable during a partition and will need to spend
+time and resources to resynchronize any missing data and rebalance itself once
+the partition heals. Recovering from a failure may involve copying a large
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:248]: It seems like we're talking around reallocation and shard recovery here. Is there a reason we don't just directly mention and xref those two concepts?

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 8]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
+If you have divided your cluster into zones then typically the network
+connections within each zone are of higher quality than the connections between
+the zones. You must make sure that the network connections between zones are of
+sufficiently high quality. You will see the best results by locating all your
+zones within a single data center with each zone having its own independent
+power supply and other supporting infrastructure. You can also _stretch_ your
+cluster across nearby data centers as long as the network interconnection
+between each pair of data centers is good enough.
+
+[[high-availability-cluster-design-min-network-perf]] There is no specific
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:261]: I'd add some whitespace for readability.
```suggestion
[[high-availability-cluster-design-min-network-perf]]
There is no specific
```

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 9]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
-length of time for a cluster to fully recover from a network partition since it
-must resynchronize any missing data and rebalance the cluster once the
-partition heals. If you want your data to be available in multiple data centres,
-deploy a separate cluster in each data centre and use
-<<modules-cross-cluster-search,{ccs}>> or <<xpack-ccr,{ccr}>> to link the
+{es} expects its node-to-node connections to be reliable and to have low
+latency and adequate bandwidth. Many of the tasks that {es} performs require
+multiple round-trips between nodes. This means that a slow or unreliable
+interconnect may have a significant effect on the performance and stability of
+your cluster. A few milliseconds of latency added to each round-trip can
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:243]: I'd consider breaking the paragraph here.
```suggestion
your cluster.

For example, a few milliseconds of latency added to each round-trip can
```

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 10]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
+quickly accumulate into a noticeable performance penalty. {es} will
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:244]: We may want to introduce network partitions and their cause after the first sentence.

```suggestion
quickly accumulate into a noticeable performance penalty. An unreliable network
may have frequent network partitions. {es} will
```

**Review by jrodewig — APPROVED** [review]
Thanks for this update @DaveCTurner. Great writing as usual. I left some non-blocking feedback that you can ignore if wanted.

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 11]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
+connections within each zone are of higher quality than the connections between
+the zones. You must make sure that the network connections between zones are of
+sufficiently high quality. You will see the best results by locating all your
+zones within a single data center with each zone having its own independent
+power supply and other supporting infrastructure. You can also _stretch_ your
+cluster across nearby data centers as long as the network interconnection
+between each pair of data centers is good enough.
+
+[[high-availability-cluster-design-min-network-perf]] There is no specific
+minimum network performance required to run a healthy {es} cluster. In theory a
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:262]:
```suggestion
minimum network performance required to run a healthy {es} cluster. In theory, a
```

File: docs/reference/high-availability/cluster-design.asciidoc [hunk 12]
```diff
@@ -230,24 +230,47 @@ The cluster will be resilient to the loss of any node as long as:
...
+cluster will work correctly even if the round-trip latency between nodes is
+several hundred milliseconds. In practice if your network is that slow then the
```
Comment (jrodewig) [inline docs/reference/high-availability/cluster-design.asciidoc:264]:
```suggestion
several hundred milliseconds. In practice, if your network is that slow then the
```

File: docs/reference/high-availability/cluster-design.asciidoc (hunk 7 above)
Comment (DaveCTurner) [reply docs/reference/high-availability/cluster-design.asciidoc]: I didn't think we had any docs on those topics, at least not concept-level ones that would be suitable for linking from here. If you have some in mind then sure we can add links.

**Comment by DaveCTurner** [comment]
Merging this in a spirit of progress over perfection, we can continue to iterate (adding links and following up on thoughts from Michael and Dan and so on) as needed.

File: docs/reference/high-availability/cluster-design.asciidoc (hunk 7 above)
Comment (jrodewig) [reply docs/reference/high-availability/cluster-design.asciidoc]: I think you're right. We have some setting reference:

- https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-cluster.html
- https://www.elastic.co/guide/en/elasticsearch/reference/master/recovery.html

However, I don't think those are great links to use here. I've opened https://github.com/elastic/elasticsearch/issues/77515 to track this gap and add those docs.

This looks fine to me in the meantime.

**Pull request merged**